### PR TITLE
show devices limit

### PR DESCRIPTION
**Pull request by haimeiguo**
now SHOW DEVICES does not support limit and offset keyword. This pr add limit offset to SHOW DEVICES

### CODE REVIEW

**Review by mychaow — APPROVED** [review]
solve the sonar problem

**Comment by HTHou** [comment]
Hi, I have just one suggestion. Can you please provide more details about what this PR trying to fix? Or create an issue?
Otherwise some people who want to know what the PR is may have to see the code...

**Review by Alima777 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java [hunk 1]
```diff
@@ -1134,16 +1160,25 @@ private void findChildNodePathInNextLevel(
...
+          if (hasLimit) {
+            curOffset.set(curOffset.get() + 1);
+            if (curOffset.get() < offset.get() || count.get().intValue() == limit.get().intValue()) {
+              return;
+            }
+          }
           res.add(node.getPartialPath());
+          if (hasLimit) {
+            count.set(count.get() + 1);
+          }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/metadata/MTree.java:1179]:
```suggestion
          if (hasLimit) {
            curOffset.set(curOffset.get() + 1);
            if (curOffset.get() < offset.get() || count.get().intValue() == limit.get().intValue()) {
              return;
            }
            count.set(count.get() + 1);
          }
          res.add(node.getPartialPath());
```

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java [hunk 2]
```diff
@@ -1155,10 +1190,19 @@ private void findDevices(MNode node, String[] nodes, int idx, Set<PartialPath> r
...
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/metadata/MTree.java:1202]: The same.

**Review by Alima777 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/qp/physical/sys/ShowDevicesPlan.java [hunk 3]
```diff
@@ -21,15 +21,49 @@
...
     super(showContentType);
     this.path = path;
   }
 
+  public ShowDevicesPlan(ShowContentType showContentType, PartialPath path, int limit, int offset) {
+    this(showContentType, path);
+    this.limit = limit;
+    this.offset = offset;
+  }
+
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/sys/ShowDevicesPlan.java:40]:
```suggestion
  public ShowDevicesPlan(ShowContentType showContentType, PartialPath path, int limit, int offset) {
    this(showContentType, path);
    this.offset = offset;
     if (limit == 0) {
          this.limit = fetchSize;
          this.hasLimit = false;
        } else {
          this.hasLimit = true;
        }
  }

```

Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/sys/ShowDevicesPlan.java:40]: Why not put the initialization here? :D

**Review by Alima777 — CHANGES_REQUESTED** [review]

**Comment by Alima777** [comment]
And one more suggestion is: since `showDevicesDataSet` is highly familiar with `showTimeseriesDataSet`, and there is lots of dupulication.  Consider abstract an `showDataSet` as a base class, the both above can extends it.

**Review by Alima777 — APPROVED** [review]
It looks good to me now~ Thank you :D

**Review by HTHou — COMMENTED** [review]
@haimeiguo 
LGTM but there are 10 code smells. I think most of them should be resolved...

File: server/src/main/java/org/apache/iotdb/db/qp/physical/sys/ShowPlan.java [hunk 4]
```diff
@@ -44,6 +65,34 @@ public ShowContentType getShowContentType() {
...
+    return offset;
+  }
+
+  public void setOffset(int offset) {
+    this.offset = offset;
+  }
+
+  public boolean hasLimit() {
+    return hasLimit;
+  }
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/sys/ShowPlan.java:90]: The sonar test shows these lines are duplicated with some lines in `ShowTimeseriesPlan`. How do you think remove the lines in `ShowTimeseriesPlan`?

Comment (haimeiguo) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/sys/ShowPlan.java:90]: code smell solved. thank you :)

**Review by HTHou — APPROVED** [review]

**Pull request merged**